### PR TITLE
Abstract engine

### PR DESCRIPTION
**Pull request by Isaac-DeFrain**
This is PR introduces the abstract engine for Mina Foundation's integration testing efforts and general improvements.

The general idea is that the abstract engine should be agnostic to the network runner. It operates by using network runner-specific config files (e.g. the mock config located in `./integration_tests/`) and simply running bash commands based on the config. The stdout of the network runner commands are captured by the abstract engine and parsed into the response types defined in `config_util.ml` which are subsequently used to manage Lucy's internal network state. Network runner responses must conform to `return_spec.json`. Node interactions still proceed via GQL.

The other major departures from the GCP engine are the abstract engine's generation of a topology file to be consumed by the network runner and use of pre-generated network and libp2p keys to boost performance.

- adds the abstract engine module in `src/lib/integration_test_abstract_engine`
- adds topology file generation
- adds use of pre-generated network and libp2p keys
- adds config and return spec

### CODE REVIEW

**Comment by Isaac-DeFrain** [comment]
You can run the unit tests in `config_util.ml` with

```sh
dune runtest src/lib/integration_test_abstract_engine/
```

and the unit tests in `test_config.ml` with

```sh
dune runtest src/lib/integration_test_lib/
```

**Comment by Isaac-DeFrain** [comment]
Check out `src/lib/integration_test_abstract_engine/unit_tests.ml` to see how things are working

**Comment by ejMina226** [comment]
@Isaac-DeFrain Is it OK to delete the mli files for the current integration tests? Have you confirmed this by still being able to run them? Thanks.

File: src/lib/integration_test_abstract_engine/terraform.ml
Comment (ejMina226) [inline src/lib/integration_test_abstract_engine/terraform.ml:1]: Do we need a terraform file in the integration_test_abstract_engine?

File: src/lib/merkle_address/merkle_address.ml
Comment (ejMina226) [inline src/lib/merkle_address/merkle_address.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/mina_base/zkapp_command.ml
Comment (ejMina226) [inline src/lib/mina_base/zkapp_command.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/mina_base/sparse_ledger_base.ml
Comment (ejMina226) [inline src/lib/mina_base/sparse_ledger_base.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/mina_incremental/mina_incremental.ml
Comment (ejMina226) [inline src/lib/mina_incremental/mina_incremental.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/network_pool/batcher.ml
Comment (ejMina226) [inline src/lib/network_pool/batcher.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/network_pool/indexed_pool.ml
Comment (ejMina226) [inline src/lib/network_pool/indexed_pool.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/network_pool/indexed_pool.mli
Comment (ejMina226) [inline src/lib/network_pool/indexed_pool.mli:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/parallel_scan/parallel_scan.ml
Comment (ejMina226) [inline src/lib/parallel_scan/parallel_scan.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/parallel_scan/parallel_scan.mli
Comment (ejMina226) [inline src/lib/parallel_scan/parallel_scan.mli:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/ppx_version/versioned_module.ml
Comment (ejMina226) [inline src/lib/ppx_version/versioned_module.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/transaction_snark/transaction_snark.ml
Comment (ejMina226) [inline src/lib/transaction_snark/transaction_snark.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/trust_system/peer_trust.ml
Comment (ejMina226) [inline src/lib/trust_system/peer_trust.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

**Comment by ejMina226** [comment]
@Isaac-DeFrain  I know this is a WIP PR, and is still subject to change, but can you please be mindful of making subjective spacing changes. The PR is already quite big and it makes it harder to see the actual code changes. Thanks.

File: src/lib/consensus/proof_of_stake.ml
Comment (ejMina226) [inline src/lib/consensus/proof_of_stake.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

Comment (ejMina226) [reply src/lib/consensus/proof_of_stake.ml:1]: Can you please revert this change? I don't think this file should be touched at all in this PR. Thanks.

File: src/lib/zkapps_examples/zkapps_examples.ml
Comment (ejMina226) [inline src/lib/zkapps_examples/zkapps_examples.ml:1]: Did you deliberately change the spacing in this file? Or is it a spacing issue with your editor?

File: src/lib/integration_test_cloud_engine/kubernetes_network.ml
Comment (ejMina226) [inline src/lib/integration_test_cloud_engine/kubernetes_network.ml:1]: I see that these files for the old GCP engine have been touched.  I assume it's because we want them to be covered by the same process that targets the new abstract engine. Can I just confirm that no config file or contract will be expected in this case? Thanks.

File: src/lib/integration_test_abstract_engine/config.json
Comment (ejMina226) [inline src/lib/integration_test_abstract_engine/config.json:1]: This looks good. Is this just an example of what the minimina config file should look like? Do we intend to store the config files for the various environments in this directory when the engine is ready? Thanks.

File: src/lib/integration_test_abstract_engine/ci_interaction.ml
Comment (ejMina226) [inline src/lib/integration_test_abstract_engine/ci_interaction.ml:1]: Could you please tell me the purpose of this file? Does it have an analogue in the GCP case or is it just the Network_config that was originally in the mina_automation.ml? Thanks.

File: src/app/test_executive/test_executive.ml [hunk 1]
```diff
@@ -44,29 +51,52 @@ let validate_inputs ~logger inputs (test_config : Test_config.t) :
...
+        Deferred.unit
 
 let engines : engine list =
-  [ ("cloud", (module Integration_test_cloud_engine : Intf.Engine.S)) ]
+  [ ("abstract", (module Integration_test_abstract_engine : Intf.Engine.S))
+  ; ("cloud", (module Integration_test_cloud_engine : Intf.Engine.S))
+  ]
 
 let tests : test list =
-  [ ( "peers-reliability"
```
Comment (ejMina226) [inline src/app/test_executive/test_executive.ml:53]: Now that the labels are gone from the test list, are we still capable of targeting a specific test, e.g. TestExecutive.execute "peers-reliability"? The TestnetIntegration.dhall  requires this.

File: src/lib/integration_test_abstract_engine/terraform.ml
Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/terraform.ml:1]: Almost certainly, we do not. Some clean up of the `integration_test_abstract_engine` dir is definitely warranted

File: src/lib/merkle_address/merkle_address.ml
Comment (Isaac-DeFrain) [reply src/lib/merkle_address/merkle_address.ml:1]: Something weird happened with `reformat`. I'll look into it and reset these files

File: src/lib/integration_test_cloud_engine/kubernetes_network.ml
Comment (Isaac-DeFrain) [reply src/lib/integration_test_cloud_engine/kubernetes_network.ml:1]: I had to change the integration test `intf.ml` so to make the GCP engine compatible, it required innocuous changes to the GCP engine code. Nothing is changed about functionality there

File: src/lib/integration_test_abstract_engine/config.json
Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/config.json:1]: This is an example config used solely for unit tests. I'm open to doing this a different way.
The config files will be stored in `integration_tests` most likely along with a doc explaining what they are used for.

File: src/lib/integration_test_abstract_engine/ci_interaction.ml
Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/ci_interaction.ml:1]: The name is a bit misleading, probably something along the lines `config_util.ml` would be more suitable. This file defines all the datatypes used in config file interactions, handles all the config file parsing, specific config command running, and handling of config command responses from the network runners. There is no analog in the GCP engine. The `Network_config` module was moved here because of dependencies, but I'm sure we can put it in it's own file.

File: src/app/test_executive/test_executive.ml (hunk 1 above)
Comment (Isaac-DeFrain) [reply src/app/test_executive/test_executive.ml:53]: Good question! Yes. The test names are declared in their respective modules and a function for extracting the name from the first-class test functor and `Inputs` module is provided via `test_name`. In other words, nothing about this functionality has changed.

File: src/lib/integration_test_abstract_engine/config.json
Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/config.json:1]: One option would be to make a `config` value in the `unit_tests.ml` file which is just the contents of this file as a string. That way we could also eliminate the file reading.

Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/config.json:1]: However, the way the `run_command` function works is it takes a `config` arg which is the path to the config file. We could add some optional arg(s) and logic to override this in the tests, but I'm not convinced it's worth the effort.

File: src/lib/integration_test_abstract_engine/terraform.ml
Comment (Isaac-DeFrain) [reply src/lib/integration_test_abstract_engine/terraform.ml:1]: Removed

File: src/lib/integration_test_abstract_engine/ci_interaction.ml
Comment (ejMina226) [reply src/lib/integration_test_abstract_engine/ci_interaction.ml:1]: Thanks for the explanation. Yes, I think renaming it would be best for clarity.

File: src/lib/integration_test_abstract_engine/config.json
Comment (ejMina226) [reply src/lib/integration_test_abstract_engine/config.json:1]: Thanks. No, keep it as a file, but please move it to another location (or sub-directory) so that it's clear this is just for testing and serves no functional value.

File: src/lib/integration_test_abstract_engine/ci_interaction.ml [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+                   (k, Network_keypair.to_yojson v) :: accum )
+                 map )] )
+    ; constants : Test_config.constants
+    ; config : config
+    }
+  [@@deriving to_yojson]
+
+  let testnet_log_filter t = t.config.network_id
+
+  (* TODO: add config file command call here *)
```
Comment (ejMina226) [inline src/lib/integration_test_abstract_engine/ci_interaction.ml]: Are you sure the expand function should call the config file? As far as I'm aware in the GCP case, the expand method is just taking the test_config defined in each of the integration tests and expanding it, e.g. creating key_pairs, constraints and genesis_ledgers. This information is then used in the create and deploy functions, which will need to call commands from the config.

File: integration_tests/return_spec.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{
+  "version": 1,
```
Comment (ejMina226) [inline integration_tests/return_spec.json:2]: Why do we need a return_spec? Thanks.

File: src/lib/integration_test_abstract_engine/abstract_network.ml [hunk 4]
```diff
@@ -0,0 +1,1292 @@
...
+  |> Core.String.Map.of_alist_exn
+
+let genesis_keypairs { genesis_keypairs; _ } = genesis_keypairs
+
+let all_node_id t =
+  let pods = all_pods t |> Core.Map.to_alist in
+  List.fold pods ~init:[] ~f:(fun acc (_, node) -> node.node_id :: acc)
+
+(* TODO: expand or remove *)
+let initialize_infra ~logger network =
```
Comment (ejMina226) [inline src/lib/integration_test_abstract_engine/abstract_network.ml]: We can't remove this as it's in the intf.ml file and is called in the `test_executive` itself. To minimise changes to the `cloud` engine we should keep the function as is below. In the GCP case, the function looks like it's just testing all the pods were deployed with nodes in them. Given that we've abstracted away the infrastructure from the test I assume this will now  be done in the network runner (e.g. minimina or AWS), so we don't need anything other than a unit returning function, but we will need some process like this in the network runner that basically checks the network is ready to start the nodes.

File: integration_tests/return_spec.json (hunk 3 above)
Comment (Isaac-DeFrain) [reply integration_tests/return_spec.json:2]: This is the interface which the network runners need to implement in order to interact with Lucy successfully

File: src/lib/integration_test_lib/intf.ml [hunk 5]
```diff
@@ -199,10 +197,10 @@ module Engine = struct
         -> pk:Signature_lib.Public_key.Compressed.t
         -> string list Deferred.Or_error.t
 
-      val get_peer_id :
+      val get_peer_ids :
         logger:Logger.t -> t -> (string * string list) Deferred.Or_error.t
```
Comment (ejMina226) [inline src/lib/integration_test_lib/intf.ml]: Hi. Can you tell me why this has been changed? If it's serving no functional purpose, it might be worth changing it back. I would like the PR to be as lean as possible with only changes relevant to the abstract engine. Thanks.

File: src/lib/integration_test_lib/intf.ml [hunk 6]
```diff
@@ -199,10 +197,10 @@ module Engine = struct
...
 
-      val must_get_peer_id :
+      val must_get_peer_ids :
         logger:Logger.t -> t -> (string * string list) Malleable_error.t
```
Comment (ejMina226) [inline src/lib/integration_test_lib/intf.ml]: Same as above.

File: src/lib/integration_test_cloud_engine/mina_automation.ml [hunk 7]
```diff
@@ -134,18 +135,20 @@ module Network_config = struct
...
-    (* GENERATE ACCOUNTS AND KEYPAIRS *)
-    let keypairs =
-      List.take
-        (* the first keypair is the genesis winner and is assumed to be untimed. Therefore dropping it, and not assigning it to any block producer *)
-        (List.drop
-           (Array.to_list (Lazy.force Key_gen.Sample_keypairs.keypairs))
-           1 )
+    (* ACCOUNTS AND KEYPAIRS *)
+    (* ACCOUNTS AND KEYPAIRS *)
+    let before = Time.now () in
```
Comment (ejMina226) [inline src/lib/integration_test_cloud_engine/mina_automation.ml]: In the interest of keeping the GCP engine as close to what it was before building the new engine, I wonder if we should leave the key generation as is (for the GCP engine) and only have the new key generation for the abstract engine. I'll raise in the stand-up.

File: src/lib/integration_test_lib/intf.ml [hunk 8]
```diff
@@ -254,6 +253,23 @@ module Engine = struct
...
+
+    (*** [config_path] is instantiated when command line args are parsed *)
+    val config_path : string ref
+
+    (*** [keypairs_path] is instantiated when command line args are parsed *)
+    val keypairs_path : string ref
+
+    (*** [mina_image] is instantiated when command line args are parsed *)
+    val mina_image : string ref
   end
```
Comment (ejMina226) [inline src/lib/integration_test_lib/intf.ml:122]: @Isaac-DeFrain Why have these been added to the intf? Weren't the archive image and mina_image already being passed into the cli for the old gcp engine?

File: src/app/test_executive/test_executive.ml [hunk 9]
```diff
@@ -437,28 +505,72 @@ let debug_arg =
...
+  let set_config path =
+    Option.iter path ~f:(fun p -> Engine.Network.config_path := p) ;
+    path
+  in
+  let set_keypair path =
+    Engine.Network.keypairs_path := path ;
+    path
+  in
+  let set_mina_image image =
+    Engine.Network.mina_image := image ;
```
Comment (ejMina226) [inline src/app/test_executive/test_executive.ml:552]: Why do we have this now? Previously GCP was using the mina_image and arcive_image, so what's changed? Thanks.

File: src/lib/integration_test_lib/util.ml [hunk 10]
```diff
@@ -9,6 +9,61 @@ let run_cmd dir prog args =
   Process.create_exn ~working_dir:dir ~prog ~args ()
   >>= Process.collect_output_and_wait
 
+let drop_outer_quotes s =
+  let open String in
```
Comment (ejMina226) [inline src/lib/integration_test_lib/util.ml:13]: Why is this needed?

File: src/lib/integration_test_lib/intf.ml [hunk 11]
```diff
@@ -263,7 +279,11 @@ module Engine = struct
 
     type t
 
-    val create : logger:Logger.t -> Network_config.t -> t Malleable_error.t
+    val create :
+         logger:Logger.t
+      -> Network_config.t
+      -> Test_config.t
+      -> t Malleable_error.t
```
Comment (ejMina226) [inline src/lib/integration_test_lib/intf.ml]: Why is Test_config.t a new argument here?

File: src/lib/integration_test_lib/util.ml (hunk 10 above)
Comment (Isaac-DeFrain) [reply src/lib/integration_test_lib/util.ml:13]: Because of the way `Yojson.Safe` serializes strings,
```
`String "hello"
```
is serialized to `"hello"`

File: src/app/test_executive/test_executive.ml (hunk 9 above)
Comment (Isaac-DeFrain) [reply src/app/test_executive/test_executive.ml:552]: Because we need this info inside the topology file which is produced in a module that `test_executive` depends on

File: src/lib/integration_test_lib/intf.ml (hunk 8 above)
Comment (Isaac-DeFrain) [reply src/lib/integration_test_lib/intf.ml:122]: These values are used in the `network_config` module, but their values are set in the `test_executive` module when the command line args are parsed. Thus, we need a way to access them in the `test_executive`, e.g. via `Engine.Network.keypairs_path`

File: src/lib/integration_test_lib/intf.ml (hunk 5 above)
Comment (Isaac-DeFrain) [reply src/lib/integration_test_lib/intf.ml]: This name is strictly more correct, but I'll change it back to minimize the number of changes to GCP related modules

File: src/lib/integration_test_lib/intf.ml (hunk 6 above)
Comment (Isaac-DeFrain) [reply src/lib/integration_test_lib/intf.ml]: Same

File: src/lib/integration_test_cloud_engine/mina_automation.ml [hunk 12]
```diff
@@ -173,7 +173,7 @@ module Network_config = struct
       | [] ->
           mp
       | hd :: tl ->
-          let ( { Test_config.Test_Account.balance; account_name; timing }
+          let ( { Test_config.Test_Account.balance; account_name; timing; _ }
               , (pk, sk) ) =
```
Comment (ejMina226) [inline src/lib/integration_test_cloud_engine/mina_automation.ml]: The `_` after `timing` is what I had to add for the code to compile when I reintroduced the key-pair generation for gcp.

File: src/lib/integration_test_lib/intf.ml (hunk 8 above)
Comment (ejMina226) [reply src/lib/integration_test_lib/intf.ml:122]: How was this working in the old gcp engine as it needed the same information?

Comment (Isaac-DeFrain) [reply src/lib/integration_test_lib/intf.ml:122]: I'm not entirely sure, but they definitely weren't using these values higher up in the dependency DAG as we are in the abstract engine

File: src/lib/integration_test_cloud_engine/mina_automation.ml (hunk 12 above)
Comment (Isaac-DeFrain) [reply src/lib/integration_test_cloud_engine/mina_automation.ml]: Now it's showing. That's fine and needed because I added `pk` and `sk` fields directly to `Test_account.t`. All testing engines use the `Test_config` module, but the GCP engine just doesn't need these fields in `Test_account.t` because they don't pre-generate the keys

**Comment by Isaac-DeFrain** [comment]
The newest incarnation of the incantation

```bash
dune exec src/app/test_executive/test_executive.exe -- abstract mock \
  --mina-image provided-mina-image \
  --archive-image provided-archive-image 
  --config ./integration_tests/config/mock.json \
  --keypairs-path ../lucy-keypairs/ \
  --debug \
| tee mina.log | logproc -i inline -f '!(.level in ["Spam", "Debug"])'
```

**Review by ejMina226 — COMMENTED** [review]
We still need to remove the code for the mock network, but this looks good.

**Comment by Isaac-DeFrain** [comment]
Do we want to merge this into `develop` or `mf-develop`?

**Comment by ejMina226** [comment]
@michal0mina Isaac's question is probably one for you.

**Review by ejMina226 — APPROVED** [review]

**Comment by kaozenn** [comment]
@ejMina226 , this repository is no longer relevant. Can we close the pull request?
